[docs] Add the subobjects adoption guide to docs (#9874)
elastic/integrations: docs/subobjects_adoption_guide.md

# Goals

> This document that summarizes the history of subobjects, the motivation for adding the option, and recommended steps to guide integration users and developers in using it.

Reviewer: ```suggestion
This document summarizes the history of subobjects, the motivation for adding the option, and recommended steps to guide integration users and developers in using it.
```

Author: Ouch, thanks for fixing this: the sentence flows much better now.

## Overview

> [subobjects](https://www.elastic.co/guide/en/elasticsearch/reference/current/subobjects.html) is a mapping parameter that controls the configuration of the object mapper (root as well as individual fields), making it possible to store logs and metrics documents that have fields with dots in their names.

Reviewer: Nit. Link to an specific version in case the link changes in future versions.
```suggestion
[subobjects](https://www.elastic.co/guide/en/elasticsearch/reference/8.13/subobjects.html) is a mapping parameter that controls the configuration of the object mapper (root as well as individual fields), making it possible to store logs and metrics documents that have fields with dots in their names.
```

Author: Great point. 8.13 was my intention, but I missed checking the URL. Thanks.

For example:

```json
{
  "metrics.time" : 10,
  "metrics.time.min" : 1,
  "metrics.time.max" : 500
}
```

This document shares how to start using subobjects in integrations.

## History

### Elasticsearch

In early spring 2022, Elasticsearch v8.3 [introduced](https://github.com/elastic/elasticsearch/pull/86166) the subobjects mapping parameter, which added support for dots in field names for metrics use cases.

The subobjects mapping parameter made it possible to store metrics data with fields in the metric names:

```json
{
  "metrics.time" : 10,
  "metrics.time.min" : 1,
  "metrics.time.max" : 500
}
```

The subobjects parameter controls whether an object can hold other objects (defaults to true) or not.

#### Use outside integrations

We started using it in support requests to address problems in the metrics space caused by dots in field names.

### Integrations

> In winter 2023, we started [adding support for the subobjects](https://github.com/elastic/package-spec/issues/349) mapping parameter to package-spec and Kibana.

Reviewer: Nit. On what hemisphere :slightly_smiling_face: 
```suggestion
By the end of 2023, we started [adding support for the subobjects](https://github.com/elastic/package-spec/issues/349) mapping parameter to package-spec and Kibana.
```

Author: LOL, thanks! I need to think in a distributed way!

#### Field vs. Data Stream

> Users can set the subobjects mapping parameter on a single field or for the whole data stream (root level).

Reviewer: Should we give any recommendation about when to use one or the other option?

Author: Yep, adding a couple of paragraphs.

Author: @jsoriano, I moved this section to the following overview; it seems to make more sense. I also added a couple of paragraphs that mention cases for field and data stream levels.

Let me know what you think.

When the last requirement was merged in Elasticsearch, we added support for per-field subobjects in stack 8.13 and per data stream in stack 8.14. 

### High-level Overview

The default value for subobjects is true, so Elasticsearch turns fields with dots in their names into an object tree.

The subobjects mapping parameter applies to single fields and the whole data stream.

#### Per field

You can start small by disabling subobjects on a single field.

```yaml
- name: tags.*
  type: object
  object_type: keyword
  object_type_mapping_type: "*"
  subobjects: false
  description: >
    Azure resource tags.
```

#### Per data stream

You can also disable subobjects at the root level on the whole data stream.

```yaml
title: Example
type: logs
elasticsearch:
  index_template:
    mappings:
      subobjects: false
```

## Availability

### Per field
#### Requirements

- Stack version: 8.13
- package-spec version: 3.1.0

For more information, see the following:
- https://github.com/elastic/kibana/pull/171826
- https://github.com/elastic/package-spec/pull/573 

#### Recommendations

> Consider upgrading to the recent package-spec according to your minimum stack requirements. The benefits (especially additional checks that elastic-package delivers) outweigh the costs.

Reviewer: :heart:

Author: It's true, learned from experience.
